### PR TITLE
Frontend part for adding Api Documentation to the services

### PR DESCRIPTION
**Pull request by couniojc**
** Describe the change **

This is the frontend part of the API Documentation feature. It optionally adds one column in the ServiceList and an 'API Doc' tab in the ServiceDetails pages

** Issue reference **

https://github.com/kiali/kiali/issues/1237

** Backwards compatible? **

[X] Is your pull-request introducing changes in behaviour?

It adds new column and new tab if annotations are set on the services

** Screenshot **

Screen Shot 2019-07-19 at 3 41 29 PM
Screen Shot 2019-07-19 at 3 40 59 PM

** Documentation **

https://github.com/kiali/kiali/pull/1266/commits/94a0441f7976b7039f1d45dd0f941bba3a173876

### CODE REVIEW

**Comment by rhqci** [comment]
Can one of the admins verify this patch?

File: package.json [hunk 1]
```diff
@@ -54,7 +54,7 @@
     "start:dev": "npm run remove-rcue; sh -ac 'source ./.env.upstream; npm run start:kiali'",
     "start:kiali": "npm run copy-fonts && npm run copy-img && npm run build-css && REACT_APP_VERSION=$npm_package_version REACT_APP_NAME=$npm_package_name REACT_APP_GIT_HASH=$(git rev-parse HEAD) react-scripts start",
     "start:prod": "npm run restore-rcue; sh -ac 'source ./.env.downstream; npm run start:kiali'",
-    "test": "npm run build-css && TEST_RUNNER=1 react-scripts test --env=jsdom __tests__",
+    "test": "npm run build-css && TEST_RUNNER=1 craco test --env=jsdom __tests__",
```
Comment (cfcosta) [inline package.json]: Why this change?

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    return style({
+      marginTop: -2,
+      marginRight: 6,
+      width: 30
+    });
+  }
+
+  render() {
+    if (this.props.apiType) {
+        if (this.props.apiType === 'grpc' || this.props.apiType === 'graphql' || this.props.apiType === 'rest') {
```
Comment (cfcosta) [inline src/components/ApiDocumentation/ApiTypeIndicator.tsx]: This should be part of `renderIcon`.

File: src/components/ApiDocumentation/__tests__/ApiTypeIndicator.test.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import { ApiTypeIndicator } from '../ApiTypeIndicator';
+
+describe('ApiTypeIndicator', () => {
+  it('renders when not known', () => {
+    let wrapper = shallow(<ApiTypeIndicator apiType="unknown" />);
+    expect(wrapper.html()).not.toContain('img');
+  });
+
+  it('renders when known', () => {
```
Comment (cfcosta) [inline src/components/ApiDocumentation/__tests__/ApiTypeIndicator.test.tsx]: The test names do not explicit what is being under test.

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx [hunk 4]
```diff
@@ -352,6 +357,9 @@ class ServiceDetails extends React.Component<ServiceDetailsProps, ServiceDetails
                     </>
                   </NavItem>
                 ))}
+              {this.state.serviceDetailsInfo.apiDocumentation.baseUrl && (
```
Comment (cfcosta) [inline src/pages/ServiceDetails/ServiceDetailsPage.tsx]: What happens here if `apiDocumentation` is null or the base url is `''`? And also, why does it get it from the state instead of using the proper props from Redux?

File: src/pages/ServiceList/ItemDescription.tsx [hunk 5]
```diff
@@ -6,20 +6,23 @@ import { DisplayMode, HealthIndicator } from '../../components/Health/HealthIndi
...
+  columnWidth: number;
 }
 
 export default class ItemDescription extends React.PureComponent<Props, State> {
   private promises = new PromisesRegistry();
 
   constructor(props: Props) {
     super(props);
-    this.state = { health: undefined };
+    this.state = { health: undefined, columnWidth: props.hasApiColumn?3:4 };
```
Comment (cfcosta) [inline src/pages/ServiceList/ItemDescription.tsx]: I don't really like the magic numbers here, specially without proper padding.

File: src/pages/ServiceList/ItemDescription.tsx [hunk 6]
```diff
@@ -51,22 +54,25 @@ export default class ItemDescription extends React.PureComponent<Props, State> {
...
+        <Col xs={12} sm={12} md={this.state.columnWidth} lg={this.state.columnWidth}>
           <strong>Config: </strong>{' '}
           <ConfigIndicator
             id={this.props.item.name + '-config-validation'}
             validations={[this.props.item.validation]}
             size="medium"
           />
         </Col>
-        <Col xs={12} sm={12} md={4} lg={4} />
+        <Col xs={12} sm={12} md={this.state.columnWidth} lg={this.state.columnWidth}>
```
Comment (cfcosta) [inline src/pages/ServiceList/ItemDescription.tsx]: Does it help about the width if the code just hide the whole `Col` if it's  not present?

File: src/pages/ServiceList/ItemDescription.tsx [hunk 7]
```diff
@@ -51,22 +54,25 @@ export default class ItemDescription extends React.PureComponent<Props, State> {
...
+          {this.props.item.apiType && <strong>Api: </strong>}
+          {this.props.item.apiType && <ApiTypeIndicator apiType={this.props.item.apiType} />}
```
Comment (cfcosta) [inline src/pages/ServiceList/ItemDescription.tsx]: And here both could be only one conditional (hopefully on the top element level).

File: src/pages/ServiceList/ServiceListComponent.tsx [hunk 8]
```diff
@@ -188,6 +189,14 @@ class ServiceListComponent extends ListComponent.Component<
     const pageStart = (this.state.pagination.page - 1) * this.state.pagination.perPage;
     let pageEnd = pageStart + this.state.pagination.perPage;
     pageEnd = pageEnd < this.state.listItems.length ? pageEnd : this.state.listItems.length;
+    let hasApiColumn = false;
+    for (let i = pageStart; i < pageEnd; i++) {
```
Comment (cfcosta) [inline src/pages/ServiceList/ServiceListComponent.tsx]: `hasApiColumn = this.state.listItems.slice(pageStart, pageEnd).some(item => item.apiType)`, maybe?

File: package.json (hunk 1 above)
Comment (couniojc) [reply package.json]: We are using (at least for this version) a webcomponent. When running the tests, the webcomponent is imported and called and it leads to a "customElements is not defined".
Using craco helps to override the import of the web element so tests can run.
Any alternate suggestion is welcomed

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx (hunk 2 above)
Comment (couniojc) [reply src/components/ApiDocumentation/ApiTypeIndicator.tsx]: I will move it

File: src/components/ApiDocumentation/__tests__/ApiTypeIndicator.test.tsx (hunk 3 above)
Comment (couniojc) [reply src/components/ApiDocumentation/__tests__/ApiTypeIndicator.test.tsx]: I will change it

File: package.json (hunk 1 above)
Comment (cfcosta) [reply package.json]: I think we do something like that for `SVGPathElement`, or `SVGElement` (not at the computer right now to look).

A reference that might help (it did when I found this problem): https://github.com/jsdom/jsdom/issues/1423

File: src/pages/ServiceList/ItemDescription.tsx (hunk 6 above)
Comment (couniojc) [reply src/pages/ServiceList/ItemDescription.tsx]: Actually there's already an empty column that doesn't show, but if something is in there,  then the total width is 16 and it breaks the columns.
To avoid to change the design for people not using the API functionality, I changed the columns to 3 3 3 3 instead of 4 4 4 (empty one being ignored). I'm opened to suggestions

File: src/pages/ServiceList/ItemDescription.tsx (hunk 7 above)
Comment (couniojc) [reply src/pages/ServiceList/ItemDescription.tsx]: I tried with one conditional, but i had an error saying I needed one parent. Can you suggest something in the code ? I know it's not great as it is

File: src/pages/ServiceList/ServiceListComponent.tsx (hunk 8 above)
Comment (couniojc) [reply src/pages/ServiceList/ServiceListComponent.tsx]: Wooo, I will try that :-) Thanks

File: src/pages/ServiceList/ItemDescription.tsx (hunk 7 above)
Comment (cfcosta) [reply src/pages/ServiceList/ItemDescription.tsx]: You can wrap the elements like this: `<>Element here</>`.

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx (hunk 4 above)
Comment (couniojc) [reply src/pages/ServiceDetails/ServiceDetailsPage.tsx]: If apiDocumentation is not defined the tab is not displayed (it is probably displayed if baseUrl is '' but that shouldn't happen)

Comment (cfcosta) [reply src/pages/ServiceDetails/ServiceDetailsPage.tsx]: If we're checking for presence on those maybe a good idea would be just to tag it as nullable, and then check on the whole object.

File: src/pages/ServiceList/ItemDescription.tsx (hunk 5 above)
Comment (couniojc) [reply src/pages/ServiceList/ItemDescription.tsx]: I will change to nbColumns and make the computing of size in the display part

File: package.json (hunk 1 above)
Comment (couniojc) [reply package.json]: Thanks, I was able to do the same so I will remove the craco override

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx (hunk 4 above)
Comment (couniojc) [reply src/pages/ServiceDetails/ServiceDetailsPage.tsx]: Ok, I will add a check on the full object

**Comment by couniojc** [comment]
@cfcosta Can you please have a look to the latest changes ? Thanks !

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+  render() {
+    if (this.props.apiType) {
+      return this.renderIcon(this.props.apiType);
+    }
+    return <span />;
+  }
+
+  renderIcon(apiType: string) {
+    if (this.props.apiType === 'grpc' || this.props.apiType === 'graphql' || this.props.apiType === 'rest') {
```
Comment (cfcosta) [inline src/components/ApiDocumentation/ApiTypeIndicator.tsx]: This is somewhat better, and given that we have the `nameToSource` to see if there's an icon, maybe we should check if the key is there instead of enumerating all the possibilities here too?

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx [hunk 10]
```diff
@@ -352,6 +357,9 @@ class ServiceDetails extends React.Component<ServiceDetailsProps, ServiceDetails
                     </>
                   </NavItem>
                 ))}
+              { this.state.serviceDetailsInfo.apiDocumentation && this.state.serviceDetailsInfo.apiDocumentation.baseUrl && (
```
Comment (cfcosta) [inline src/pages/ServiceDetails/ServiceDetailsPage.tsx]: baseUrl should be not `''` and not `undefined`, empty strings are not false in javascript, as far as I know, and if they are, that's black magic :P

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx [hunk 11]
```diff
@@ -395,6 +403,14 @@ class ServiceDetails extends React.Component<ServiceDetailsProps, ServiceDetails
                   />
                 </TabPane>
               )}
+              {this.state.serviceDetailsInfo.apiDocumentation && this.state.serviceDetailsInfo.apiDocumentation.baseUrl && (
```
Comment (cfcosta) [inline src/pages/ServiceDetails/ServiceDetailsPage.tsx]: Same here!

File: src/pages/ServiceList/ItemDescription.tsx [hunk 12]
```diff
@@ -49,24 +52,33 @@ export default class ItemDescription extends React.PureComponent<Props, State> {
   }
 
   render() {
+    var columnSize = 12 / this.state.nbColumns;
```
Comment (cfcosta) [inline src/pages/ServiceList/ItemDescription.tsx]: This is a lot more readable than before! :D

File: src/pages/ServiceList/ServiceListComponent.tsx [hunk 13]
```diff
@@ -188,6 +189,7 @@ class ServiceListComponent extends ListComponent.Component<
     const pageStart = (this.state.pagination.page - 1) * this.state.pagination.perPage;
     let pageEnd = pageStart + this.state.pagination.perPage;
     pageEnd = pageEnd < this.state.listItems.length ? pageEnd : this.state.listItems.length;
+    let hasApiColumn = this.state.listItems.slice(pageStart, pageEnd).some(item => item.apiType)
```
Comment (cfcosta) [inline src/pages/ServiceList/ServiceListComponent.tsx]: Nice that it worked :D

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx (hunk 9 above)
Comment (couniojc) [reply src/components/ApiDocumentation/ApiTypeIndicator.tsx]: Good point, I fixed it

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  iconStyle() {
+    return style({
+      marginTop: -2,
+      marginRight: 6,
+      width: 30
+    });
+  }
+
+  render() {
+    if (this.props.apiType) {
```
[the remark this answers is not in the document]
Comment (couniojc) [reply src/components/ApiDocumentation/ApiTypeIndicator.tsx]: Changed to ternary

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+const apiGraphqlIcon = require('../../assets/img/api-graphql-logo.svg');
+
+const nameToSource = new Map<string, string>([
+  ['grpc', apiGrpcIcon],
+  ['rest', apiRestIcon],
+  ['graphql', apiGraphqlIcon],
+]);
+
+export class ApiTypeIndicator extends React.Component<Props> {
+  iconStyle() {
```
[the remark this answers is not in the document]
Comment (couniojc) [reply src/components/ApiDocumentation/ApiTypeIndicator.tsx]: Fixed, thanks !

File: src/components/ApiDocumentation/ApiTypeIndicator.tsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
+import * as React from 'react';
+import { style } from 'typestyle';
+
+interface Props {
+  apiType: string;
+}
+
+const apiGrpcIcon = require('../../assets/img/api-grpc-logo.svg');
+const apiRestIcon = require('../../assets/img/api-rest-logo.svg');
+const apiGraphqlIcon = require('../../assets/img/api-graphql-logo.svg');
```
[the remark this answers is not in the document]
Comment (couniojc) [reply src/components/ApiDocumentation/ApiTypeIndicator.tsx]: Fixed, thanks !

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx (hunk 10 above)
Comment (couniojc) [reply src/pages/ServiceDetails/ServiceDetailsPage.tsx]: Let's fight black magic then, I added !== ''

File: src/pages/ServiceDetails/ServiceDetailsPage.tsx (hunk 11 above)
Comment (couniojc) [reply src/pages/ServiceDetails/ServiceDetailsPage.tsx]: Fixed as well

**Comment by couniojc** [comment]
@cfcosta Thanks for the review ! Super useful ! I committed latest fixes

File: src/pages/ServiceList/ItemDescription.tsx [hunk 17]
```diff
@@ -49,24 +52,33 @@ export default class ItemDescription extends React.PureComponent<Props, State> {
...
           <strong>Config: </strong>{' '}
           <ConfigIndicator
             id={this.props.item.name + '-config-validation'}
             validations={[this.props.item.validation]}
             size="medium"
           />
         </Col>
-        <Col xs={12} sm={12} md={4} lg={4} />
+        <Col xs={12} sm={12} md={columnSize} lg={columnSize}>
+          // If no api, the column is empty and not displayed
```
Comment (jotak) [inline src/pages/ServiceList/ItemDescription.tsx]: these comments are displayed on screen :)
they should either be moved inside the `{...}` block or commented as html `<!-- -->`

Comment (couniojc) [reply src/pages/ServiceList/ItemDescription.tsx]: Wooo, how come I don't see it ? Let me change it

Comment (couniojc) [reply src/pages/ServiceList/ItemDescription.tsx]: Fixed

**Comment by jotak** [comment]
I'm rebasing this PR and, i haven't seen it coming, the variable number of columns is not playing well with @aljesusg ' refactoring and virtual list. Don't worry we'll get that in.

**Comment by jotak** [comment]
superseded by https://github.com/kiali/kiali-ui/pull/1356
